scene/animation/animation_node_state_machine.cpp

```diff
@@ -254,10 +254,10 @@ void AnimationNodeStateMachinePlayback::_set_grouped(bool p_is_grouped) {
 	is_grouped = p_is_grouped;
 }
 
-void AnimationNodeStateMachinePlayback::travel(const StringName &p_state, bool p_reset_on_teleport) {
+void AnimationNodeStateMachinePlayback::travel(const StringName &p_state, bool p_reset_on_teleport, bool p_force_jump) {
```

Comment:
```suggestion
void AnimationNodeStateMachinePlayback::travel(const StringName &p_state, bool p_reset_on_teleport) {
```

Since you pointed out the need for p_force_jump, I will reconsider and make some changes regarding my comments above.

> - Add queue_travel(String p_state)
>  - Concatenates the A* path from the last element of the travel_path array to p_state. If travel_path is empty, it is equivalent to travel().

> If travel_path is empty, it is equivalent to travel().
This should not be equivalent, but should be separated as follows:

If travel_path does not exist:
- `queue_travel()` uses default_transition for the first movement
- `travel()` uses `start()` for the first movement

This makes `p_reset_on_teleport` retain its meaning as an argument to be passed to `start()`.

Reply:
I'm totally down for adding a separate function (`jump`, perhaps) instead of a `force_jump` flag on the current `travel` call. However, I don't think `queue_travel` is the right name for it, as I would expect a function by that name to append a new series of travel request to the path as opposed to replacing all pending travel requests with a new shortest path to the requested state.

The current behavior of `force_jump` does queue a travel move after the current transition, but it doesn't have the same behavior as something like an `enqueue` or `push_back`. It still replaces the entire travel path, even though it retains any transition we are already in the middle of.

Reply:
So, my suggestion is `start()` and `travel()` as functions to replace path, and `queue()` and `queue_travel()` as functions to add or alter the path. If your implementation of jump does not use travel_path, then the function name may be odd because it does not represent an actual behavior.

However, as I suggested above, I have made a final suggestion that the jump be determined by moving unconnected states within the travel_path (allowing arbitrary editing of the travel_path) rather than by the jump flag. I think implementing that is the most consistent way to avoid creating duplicate APIs.

Reply:
My implementation only involves the suggestions relating to start() and travel(). Additional behavior such as queuing an additional transition at the end of the current travel_path are out of scope for this particular feature proposal.

The original implementation is as follows:

- `travel(State destination, bool reset_on_teleport)`
  - Use A* on explicitly provided transitions to find a path to the destination state.
  - If a path is available:
    - Delete the previous travel path and replace it with the new shortest path.
  - If no path is available:
    - Teleport to the destination state.

The current implementation is as follows:

- `travel(State destination, bool reset_on_teleport, bool force_jump)`
  - If the force_jump flag is set, skip to "no path is available".
  - Use A* on explicitly provided transitions to find a path to the destination state.
  - Delete the previous travel path.
  - If a path is available:
    - Assign the new path to `path`.
  - If no path is available:
    - Add the destination state directly to the path. (The `_find_next` function has been altered to provide the details of the default transition when an explicitly defined transition is not available.)

My preferred implementation is as follows:

# Note that the `reset_on_teleport` parameter is removed as we no longer teleport when we can't find a path

- `travel(State destination)`
  - Use A* on explicitly provided transitions to find a path to the destination state.
  - Delete the previous travel path.
  - If a path is available:
    - Assign the new path to `path`.
  - If no path is available:
    - Add the destination state directly to the path. (The `_find_next` function has been altered to provide the details of the default transition when an explicitly defined transition is not available.)

- `jump(State destination)`
  - Delete the previous travel path.
  - Add the destination state directly to the path. (The `_find_next` function has been altered to provide the details of the default transition when an explicitly defined transition is not available.)


My interpretation of your suggestions are as follows:

- `start(State destination, bool reset_on_teleport)`
  - Use A* on explicitly provided transitions to find a path to the destination state.
  - Delete the previous travel path.
  - If a path is available:
    -  Assign the new path to `path`.
  - If no path is available:
    - Teleport to the destination state.

- `travel(State destination, bool reset_on_teleport)`
  - Use A* on explicitly provided transitions to find a path to the destination state.
  - Delete the previous travel path.
  - If a path is available:
    -  Assign the new path to `path`.
  - If no path is available:
    - Add the destination state directly to the path. (The `_find_next` function has been altered to provide the details of the default transition when an explicitly defined transition is not available.)

- `queue_start(State destination, bool reset_on_teleport)`
  - Use A* on explicitly provided transitions to find a path to the destination state.
  - If a path is available:
    -  Append the new path to `path`.
  - If no path is available:
    - Undefined behavior. Teleports cannot be queued as they are not transitions.

- `queue_travel(State destination, bool reset_on_teleport)`
  - Use A* on explicitly provided transitions to find a path to the destination state.
  - If a path is available:
    -  Append the new path to `path`.
  - If no path is available:
    - Append the destination state to `path`. (The `_find_next` function has been altered to provide the details of the default transition when an explicitly defined transition is not available.)

Reply:
After re-reading your suggestions multiple times to try and get it in my brain properly, I'm pretty sure the only actual disagreement here is the naming convention? "travel" and "jump" vs "travel" and "start"

Reply:
If you put the queue out of scope, we add an API with redundant arguments, but it will be difficult to remove those arguments when we add the queue later to keep compatibility, so I recommend that you add the queue and remove those arguments in this PR.

If that seems difficult for you, I would like to take over/salvage your work later by 4.5.

Reply:
Which arguments are redundant? My main proposal has just one argument (the desired state). Actually, my main proposal already breaks compatibility. Should I retain the function signatures as best I can? Or should this be considered a breaking change?

I will not put the queue out of scope as long as there is no need for an additional feature proposal. The change is not trivial, but not out of reach either.

Do you take any issue with the vocabulary `travel`, `jump`, `queue_travel`, and `queue_jump`? I would not like to use the verb `start` as it would not be related to the internal `_start` and that caused me to misunderstand your suggestions

Reply:
We do not explicitly use the word `jump`.

I assume you are indicating the behavior of using default_transiton between unconnected states. However, my suggestion is that they are automatically processed if they are queued without jump flag, so all the user needs to do is the action of queuing the unconnected states.

Reply:
> I assume you are indicating the behavior of using default_transiton between unconnected states. However, my suggestion is that they are automatically (without jump flag) processed if the unconnected states are queued, so all the user needs to do is the action of queuing the unconnected states.

Yes, that is the behavior I implemented in `travel`.

I also additionally implemented a way for the user to explicitly request the state machine to ignore existing connections, as a need for such behavior was brought to my attention [here](https://github.com/godotengine/godot/issues/98617#issuecomment-2464858233). I am aware that other solutions to this problem exist, but the use case is palpable.

Reply:
If there is a connection to the queued state, default_transition is not used, but the existing connection's transition is used.

In other words, it is not always the case that a queued connection will jump. In other words, whether or not a queued state will jump depends on whether or not that state has a connection, not on the jump request, so only `queue()` is needed here, so `jump()` and `queue_jump()` are not needed.

Reply:
https://github.com/godotengine/godot/issues/98617#issuecomment-2464417352

If you want to do this, for example, if the current is `"A"` and there is a queue created by travel `["B", "C", "D", "E"]` and you want to jump to `"E"`, just do like `clear_travel_path(); queue("E");`.

Reply:
That's actually... exactly what queue_jump does. All of the behavior you're requesting is already present, I just called it `jump` and `queue_jump` instead of just `queue`. Are you sure the vocabulary is unacceptable? It seems like a provocative verb for a common use case.

Reply:
If you understand my suggestion, the problem is that there are cases where a jump does not occur due to queue_jump. If there is a connection to the queued one it should not jump.

Reply:
> If you understand my suggestion, the problem is that there are cases where a jump does not occur by queue_jump(). If there is a connection to the queued one it should not jump (it means using default_transition) but use the one's transition.

Yes, that is the behavior. What I have been *calling* `queue_jump` does exactly what your `queue` does. The fact that I felt the need to specify "using the state machine's default transition if a direct connection is not available" in the docs points to `queue` being the more accurate term here

Reply:
I am also wondering about the difference between current `jump()` and `queue_jump()`. Perhaps `jump("State")` is synonymous with the following:

```
StateMachinePlayback.clear_travel_path()
StateMachinePlayback.queue("State")
```

and can be replaced with an implementation of `clear_travel_path()`?

Reply:
> I am also wondering about the difference between current `jump()` and `queue_jump()`. Perhaps `jump("State")` is synonymous with the following:
> 
> ```
> StateMachinePlayback.clear_travel_path()
> StateMachinePlayback.queue("State")
> ```
> 
> and can be replaced with an implementation of `clear_travel_path()`?

There is actually already a public member `clear_path`, so I don't need to implement it again. I am actually rather fond of the `jump` function for the provocative verb, even if it's just bundling up `clear_path` and `queue` as you say.

Reply:
Well, then I think `jump()` should be removed. If it is really necessary, it can be written in 3 lines on the user's side.